components/brave_wallet/browser/zcash/create_shield_all_transaction_task.cc

```diff
@@ -0,0 +1,251 @@
+  }
+
+  if (!tree_state_) {
+    GetTreeState();
+    return;
+  }
+
+  if (!utxo_map_) {
+    GetAllUtxos();
+    return;
+  }
+
+  if (!chain_height_) {
+    GetChainHeight();
+    return;
+  }
+
+  if (!transaction_ && !CreateTransaction()) {
+    std::move(callback_).Run(base::unexpected(error_.value()));
+    return;
+  }
+
+  // Signing of transparent part is final step
+  if (!transaction_->IsTransparentPartSigned()) {
+    CompleteTransaction();
+    return;
+  }
+
+  std::move(callback_).Run(std::move(transaction_.value()));
+}
+
+bool CreateShieldAllTransactionTask::CreateTransaction() {
+  CHECK(utxo_map_);
+  CHECK(chain_height_);
+
+  ZCashTransaction zcash_transaction;
+
+  // Pick inputs
+  std::vector<ZCashTransaction::TxInput> all_inputs;
+  for (const auto& item : utxo_map_.value()) {
+    for (const auto& utxo : item.second) {
+      if (!utxo) {
+        error_ = l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR);
+        return false;
+      }
+      if (auto input =
+              ZCashTransaction::TxInput::FromRpcUtxo(item.first, *utxo)) {
+        all_inputs.emplace_back(std::move(*input));
+      }
+    }
+  }
+
+  zcash_transaction.transparent_part().inputs = std::move(all_inputs);
+  // TODO(cypt4): Calculate orchard actions count
+  zcash_transaction.set_fee(CalculateZCashTxFee(
+      zcash_transaction.transparent_part().inputs.size(),
+      2 /* actions count for 1 orchard output no orchard inputs */));
+
+  // Pick orchard outputs
+  ZCashTransaction::OrchardOutput orchard_output;
+  auto addr_bytes =
+      zcash_wallet_service_->keyring_service()->GetOrchardRawBytes(
+          account_id_, mojom::ZCashKeyId::New(account_id_->account_index,
+                                              1 /* internal */, 0));
+  if (!addr_bytes) {
+    error_ = l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR);
+    return false;
+  }
+
+  if (zcash_transaction.fee() > zcash_transaction.TotalInputsAmount()) {
+    error_ = l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR);
+    return false;
+  }
+
+  orchard_output.value =
+      zcash_transaction.TotalInputsAmount() - zcash_transaction.fee();
+  orchard_output.address = std::move(addr_bytes.value());
+
+  zcash_transaction.orchard_part().outputs.push_back(std::move(orchard_output));
+
+  zcash_transaction.set_locktime(chain_height_.value());
+  zcash_transaction.set_expiry_height(chain_height_.value() +
+                                      kDefaultZCashBlockHeightDelta);
+
+  transaction_ = std::move(zcash_transaction);
+
+  return true;
+}
+
+void CreateShieldAllTransactionTask::CompleteTransaction() {
+  CHECK(transaction_);
+  CHECK_EQ(1u, transaction_->orchard_part().outputs.size());
+  CHECK(tree_state_);
+
+  // Sign shielded part
+  auto state_tree_bytes = PrefixedHexStringToBytes(
+      base::StrCat({"0x", tree_state_.value()->orchardTree}));
+  if (!state_tree_bytes) {
+    error_ = l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR);
+    Iterate();
```

Comment:
Let's do this through PostTask to avoid Iterate()->CompleteTransaction()->Iterate() recursion

Reply:
Done